drivers/pinctrl/pinctrl_stm32.c

```diff
@@ -0,0 +1,588 @@
+	case PINCTRL_STM32_FUNCTION_ALT_3:
+	case PINCTRL_STM32_FUNCTION_ALT_4:
+	case PINCTRL_STM32_FUNCTION_ALT_5:
+	case PINCTRL_STM32_FUNCTION_ALT_6:
+	case PINCTRL_STM32_FUNCTION_ALT_7:
+	case PINCTRL_STM32_FUNCTION_ALT_8:
+	case PINCTRL_STM32_FUNCTION_ALT_9:
+	case PINCTRL_STM32_FUNCTION_ALT_10:
+	case PINCTRL_STM32_FUNCTION_ALT_11:
+	case PINCTRL_STM32_FUNCTION_ALT_12:
+	case PINCTRL_STM32_FUNCTION_ALT_13:
+	case PINCTRL_STM32_FUNCTION_ALT_14:
+	case PINCTRL_STM32_FUNCTION_ALT_15:
+		return LL_GPIO_MODE_ALTERNATE;
+	case PINCTRL_STM32_FUNCTION_INPUT:
+		return LL_GPIO_MODE_INPUT;
+	case PINCTRL_STM32_FUNCTION_OUTPUT:
+		return LL_GPIO_MODE_OUTPUT;
+	case PINCTRL_STM32_FUNCTION_ANALOG:
+		return LL_GPIO_MODE_ANALOG;
+	}
+	__ASSERT(0, "function unknown - no LL_GPIO_MODE");
+	return 0;
+}
+
+static inline u32_t pinctrl_stm32_ll_gpio_alt_func(u16_t func)
+{
+	if ((func >= PINCTRL_STM32_FUNCTION_ALT_0)
+	    && (func <= PINCTRL_STM32_FUNCTION_ALT_15)) {
+		return func - PINCTRL_STM32_FUNCTION_ALT_0;
+	}
+	__ASSERT(0, "function unknown - no LL_GPIO_ALT_FUNC");
+	return func;
+}
+
+/**
+ * @brief Enable GPIO port to be clocked by associated peripherial bus clock.
+ */
+static void pinctrl_stm32_gpio_enable_clock(int port_idx)
+{
+	__ASSERT((port_idx < PINCTRL_STM32_GPIO_COUNT),
+		 "port unknown - invalid port_idx");
+
+        /* pinctrl_stm32_data.pclken_gpio.bus already set by init */
+	pinctrl_stm32_data.pclken_gpio.enr = 
+		pinctrl_stm32_gpio_data[port_idx].ll_gpio_pclken_enr;
+	clock_control_on(pinctrl_stm32_data.clk,
+                     (clock_control_subsys_t )&pinctrl_stm32_data.pclken_gpio);
+}
+
+/* --- API (using the PINCTRL lib) --- */
+
+/* forward declarations */
+static int pinctrl_stm32_config_get(struct device *dev, u16_t pin, 
+				    u32_t *config);
+static int pinctrl_stm32_config_set(struct device *dev, u16_t pin, 
+				    u32_t config);
+static int pinctrl_stm32_mux_get(struct device *dev, u16_t pin, u16_t *func);
+static int pinctrl_stm32_mux_set(struct device *dev, u16_t pin, u16_t func);
+static int pinctrl_stm32_device_init(struct device *dev);
+
+
+/* Configure PINCTRL library */
+#define CONFIG_PINCTRL_LIB_SELECT_PINMUX 1
+#define CONFIG_PINCTRL_LIB_DEVICE_NAME_PREFIX pinctrl_stm32
+#define CONFIG_PINCTRL_LIB_CONFIG_GET pinctrl_stm32_config_get
+#define CONFIG_PINCTRL_LIB_CONFIG_SET pinctrl_stm32_config_set
+#define CONFIG_PINCTRL_LIB_MUX_GET pinctrl_stm32_mux_get
+#define CONFIG_PINCTRL_LIB_MUX_SET pinctrl_stm32_mux_set
+#define CONFIG_PINCTRL_LIB_PINCONF_PINMUX_PIN PINCTRL_STM32_PINCONF_PINMUX_PIN
+#define CONFIG_PINCTRL_LIB_PINCONF_PINMUX_MUX PINCTRL_STM32_PINCONF_PINMUX_MUX
+#define CONFIG_PINCTRL_LIB_CONTROLLER_COUNT 1
+#define CONFIG_PINCTRL_LIB_CONTROLLER_0_PREFIX PINCTRL_STM32_PREFIX
+#define CONFIG_PINCTRL_LIB_CONTROLLER_0_DATA &pinctrl_stm32_data
+#define CONFIG_PINCTRL_LIB_CONTROLLER_0_PIN_COUNT PINCTRL_STM32_PIN_COUNT
+#define CONFIG_PINCTRL_LIB_CONTROLLER_0_DEVICE_INIT \
+	pinctrl_stm32_device_init
```

Comment:
while it seems a good idea, using defines to pass functions looks ... ugly

Reply:
@superna9999 Do you have a good option?